src/main/java/org/example/learningsystem/core/destination/service/DestinationServiceImpl.java

```diff
@@ -0,0 +1,79 @@
+package org.example.learningsystem.core.destination.service;
+
+import lombok.RequiredArgsConstructor;
+import org.example.learningsystem.core.destination.config.DestinationServiceProperties;
+import org.example.learningsystem.core.destination.dto.AccessTokenResponseDto;
+import org.example.learningsystem.core.destination.dto.SmtpDestinationResponseDto;
+import org.example.learningsystem.email.config.EmailServerProperties;
+import org.springframework.context.annotation.Profile;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Service;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.HttpStatusCodeException;
+import org.springframework.web.client.RestClient;
+
+import java.util.Map;
+
+import static java.util.Objects.nonNull;
+
+@Service
+@RequiredArgsConstructor
+@Profile("cloud")
+public class DestinationServiceImpl implements DestinationService {
+
+    private final DestinationServiceProperties destinationServiceProperties;
+    private final RestClient restClient = RestClient.create();
+    private static final String ACCESS_TOKEN_URI = "%s/oauth/token";
+    private static final String DESTINATION_URI = "%s/destination-configuration/v1/instanceDestinations/%s";
+    private static final String SMTP_DESTINATION = "smtp-destination";
+
+    @Retryable(retryFor = HttpStatusCodeException.class)
+    public EmailServerProperties getEmailServerProperties() {
+        var smtpDestination = tryGetDestination(SMTP_DESTINATION, SmtpDestinationResponseDto.class);
+
+        return new EmailServerProperties(smtpDestination.from(), smtpDestination.user(), smtpDestination.password(),
+                smtpDestination.host(), smtpDestination.port(), smtpDestination.protocol(),
+                smtpDestination.auth(), smtpDestination.startTls());
+    }
+
+    private <T> T tryGetDestination(String destinationName, Class<T> responseType) {
+        var uri = DESTINATION_URI.formatted(destinationServiceProperties.getUri(), destinationName);
+        var accessToken = getAccessToken();
```

Comment:
with current implementation, you will get token every request, what is not allowed with oauth2 as token has some expiration time during which you can reuse it